File: backend/app/schemas/prompt_pipeline.py
from typing import Literal

from pydantic import BaseModel, Field


PromptIntent = Literal[
    "select",
    "count",
    "aggregation",
    "ranking",
    "trend",
    "comparison",
    "expiry_monitoring",
    "blocked",
    "unknown",
]


class PromptPipelineRequest(BaseModel):
    prompt: str = Field(
        min_length=2,
        max_length=2000,
    )

    data_source_id: int | None = None

    maximum_results: int = Field(
        default=5,
        ge=1,
        le=20,
    )


class ExtractedEntityResponse(BaseModel):
    value: str
    entity_type: str
    confidence: int


class ResolvedColumnResponse(BaseModel):
    id: int
    column_name: str
    business_name: str | None
    description: str | None
    data_type: str
    classification: str
    is_sensitive: bool
    ai_access_allowed: bool

    score: int
    confidence: int
    matched_terms: list[str]
    reasons: list[str]


class ResolvedTableResponse(BaseModel):
    id: int
    data_source_id: int
    schema_name: str | None
    table_name: str
    business_name: str | None
    description: str | None
    department: str | None
    data_owner: str | None
    classification: str
    definition_status: str
    ai_access_allowed: bool

    score: int
    confidence: int
    matched_terms: list[str]
    reasons: list[str]

    columns: list[ResolvedColumnResponse]


class PromptPipelineResponse(BaseModel):
    prompt: str
    normalized_prompt: str

    intent: PromptIntent
    intent_confidence: int

    entities: list[ExtractedEntityResponse]
    keywords: list[str]
    time_expressions: list[str]
    status_terms: list[str]
    aggregation_terms: list[str]

    is_safe: bool
    blocked_reason: str | None

    metadata_confidence: int
    overall_confidence: int

    matched_tables: list[ResolvedTableResponse]

    warnings: list[str]
    explanation: list[str]

File: frontend/src/types/promptPipeline.ts
export type PromptIntent =
  | "select"
  | "count"
  | "aggregation"
  | "ranking"
  | "trend"
  | "comparison"
  | "expiry_monitoring"
  | "blocked"
  | "unknown";

export interface ExtractedEntity {
  value: string;
  entity_type:
    | "business_term"
    | "time"
    | "status"
    | "aggregation"
    | string;
  confidence: number;
}

export interface ResolvedColumn {
  id: number;
  column_name: string;
  business_name: string | null;
  description: string | null;
  data_type: string;
  classification: string;
  is_sensitive: boolean;
  ai_access_allowed: boolean;
  score: number;
  confidence: number;
  matched_terms: string[];
  reasons: string[];
}

export interface ResolvedTable {
  id: number;
  data_source_id: number;
  schema_name: string | null;
  table_name: string;
  business_name: string | null;
  description: string | null;
  department: string | null;
  data_owner: string | null;
  classification: string;
  definition_status: string;
  ai_access_allowed: boolean;
  score: number;
  confidence: number;
  matched_terms: string[];
  reasons: string[];
  columns: ResolvedColumn[];
}

export interface PromptPipelineResponse {
  prompt: string;
  normalized_prompt: string;

  intent: PromptIntent;
  intent_confidence: number;

  entities: ExtractedEntity[];
  keywords: string[];
  time_expressions: string[];
  status_terms: string[];
  aggregation_terms: string[];

  is_safe: boolean;
  blocked_reason: string | null;

  metadata_confidence: number;
  overall_confidence: number;

  matched_tables: ResolvedTable[];

  warnings: string[];
  explanation: string[];
}
